systray_linux.go

```diff
@@ -18,3 +62,268 @@ func (item *MenuItem) SetIcon(iconBytes []byte) {
+<!--  The following items are not supported by specs, but widely used  -->
+<signal name="NewIconThemePath">
+<arg type="s" name="icon_theme_path" direction="out"/>
+</signal>
+<signal name="NewMenu"/>
+<!--  ayatana labels  -->
+<!--  These are commented out because GDBusProxy would otherwise require them,
+         but they are not available for KDE indicators
+     -->
+<!-- <signal name="XAyatanaNewLabel">
+        <arg type="s" name="label" direction="out" />
+        <arg type="s" name="guide" direction="out" />
+    </signal>
+    <property name="XAyatanaLabel" type="s" access="read" />
+    <property name="XAyatanaLabelGuide" type="s" access="read" /> -->
+</interface>`+ introspect.IntrospectDataString+prop.IntrospectDataString+"</node>")
+
+	conn, _ := dbus.ConnectSessionBus()
+	conn.Export(introspect.Introspectable(intro), "/StatusNotifierItem",
+		"org.freedesktop.DBus.Introspectable")
+
+	_, err := prop.Export(conn, "/StatusNotifierItem", createPropSpec())
+	if err != nil {
+		log.Printf("Failed to export notifier item properties to bus")
+		return
+	}
+	err = conn.Export(&tray{}, "/StatusNotifierItem", "org.kde.StatusNotifierItem")
+	if err != nil {
+		log.Printf("Failed to export notifier item to bus")
+		return
+	}
+
+	name := fmt.Sprintf("org.kde.StatusNotifierItem-%d-1", os.Getpid()) // register id 1 for this process
+	_, err = conn.RequestName(name, dbus.NameFlagDoNotQueue)
+	if err != nil {
+		// fall back to existing name
+		name = conn.Names()[0]
+	}
+
+	obj := conn.Object("org.kde.StatusNotifierWatcher", dbus.ObjectPath("/StatusNotifierWatcher"))
+	call := obj.Call("org.kde.StatusNotifierWatcher.RegisterStatusNotifierItem", 0, name)
+	if call.Err != nil {
+		log.Printf("Failed to register our icon with the notifier watcher, maybe no tray running?")
+	}
+}
+
+// tray is a basic type that handles the dbus functionality
+type tray struct {
+	prop.Properties
+}
+
+// ContextMenu method is called when the user has right-clicked on our icon.
+func (t *tray) ContextMenu(x, y int) *dbus.Error {
+	// not supported for systray lib
+	return nil
+}
+
+// Activate requests that we perform the primary action, such as showing a menu.
+func (t *tray) Activate(x, y int) *dbus.Error {
+	// TODO show menu, or have it handled in the dbus?
+	return nil
+}
+
+// SecondaryActivate is alternative non-context click, such as middle mouse button.
+func (t *tray) SecondaryActivate(x, y int) *dbus.Error {
+	return nil
+}
+
+// Scroll is called when the mouse wheel scrolls over the icon.
+func (t *tray) Scroll(delta int, orient string) *dbus.Error {
+	return nil
+}
+
+type PX struct {
+	W, H int
+	Pix  []byte
+}
+
+func convertToPixels(data []byte) PX {
+	if len(iconData) == 0 {
+		return PX{}
+	}
+
+	p, _, err := image.Decode(bytes.NewReader(iconData))
+	if err != nil {
+		log.Printf("Failed to read icon format %v", err)
+		return PX{}
+	}
+
+	w, h := p.Bounds().Dx(), p.Bounds().Dy()
+	data2 := make([]byte, w*h*4)
+	i := 0
+	for y := 0; y < h; y++ {
```

Comment:
This really doesn't look great. I think the right way in Go would be to use the draw package to output to a known RGBA typed image and then return rgba.Pix ( https://pkg.go.dev/image#RGBA ).

Reply:
Sadly the output format needs to be ARGB not RGBA and none of the Go standard image formats support this to my knowledge.
Is there some way I can make it look less bad within this constraint?

Reply:
Arg, I didn't realize that. Quite annoying that Go doesn't come with the infrastructure to do ARGB nicely ( https://github.com/golang/go/issues/12966 ). A potential cleaner solution would be to use https://pkg.go.dev/golang.org/x/exp/shiny@v0.0.0-20220218215828-6cf2b201936e/driver/internal/swizzle as pointed by the comment in that issues.

Reply:
That is an alternative. However we do have to decode the image first (to get the pixel data and the X/Y size) so is the code above really an overhead?

Reply:
Updated to be more readable